### PR TITLE
Change validation error messages

### PR DESCRIPTION
**Pull request by a-alle**
- Adds custom error codes for validation errors -> only for internal usage, not user facing
- Validation throws array of errors instead of error messages concatenated with new line
- Modifies GraphQL errors thrown at validation stage:
  * strips-out any reference to Authorization types that the generated schema does not contain
  * strips-out references to the dummy values introduced to replace the "$jwt." values during the validation stage
  * adds a path argument **for custom rules only** (DirectiveArgumentOfCorrectType)
- Updates `@authentication` to be a dynamic directive -> uses an Enricher to enrich the validation schema with the `@authentication` directive on types/ fields
- Renames the `jwtPayload` argument of `@authentication` to `jwt`
- Refactors enrichers logic

### CODE REVIEW

File: packages/graphql/src/schema/validation/enrichers/directive/utils.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    InterfaceTypeExtensionNode,
+    GraphQLSchema,
+} from "graphql";
+
+export type ObjectOrInterfaceDefinitionNode = ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode;
+export type ObjectOrInterfaceExtensionNode = ObjectTypeExtensionNode | InterfaceTypeExtensionNode;
+type ObjectLikeDefinitionNode = ObjectOrInterfaceDefinitionNode | ObjectOrInterfaceExtensionNode;
+
+export type DIRECTIVE_TRANSFORM_FN = (currentDirectiveDirective: any, typeName: string) => any;
+export type CREATE_DIRECTIVE_DEFINITION_FN = (typeDefinitionName: string, schema: GraphQLSchema) => any;
```
Comment (darrellwarde) [inline packages/graphql/src/schema/validation/enrichers/directive/utils.ts:34]: What are these types? Should they live here?

File: packages/graphql/src/schema/validation/validate-sdl.ts [hunk 2]
```diff
@@ -31,7 +31,22 @@ export function validateSDL(
...
+        //     // error.name,
+        //     JSON.stringify(error.extensions, null, 2),
+        //     error.nodes,
+        //     // error.originalError,
+        //     error.stack,
+        //     error.path,
+        //     error.source,
+        //     error.positions,
+        //     error.message
+        // );
```
Comment (darrellwarde) [inline packages/graphql/src/schema/validation/validate-sdl.ts:46]: 👀

File: packages/graphql/src/schema/validation/utils/map-error.test.ts [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+                'Invalid argument: filter, error: Value "CREATE" does not exist in "AuthorizationValidateOperation" enum.',
+                {
+                    nodes: [argumentNode, directiveNode],
+                    extensions: { exception: { code: VALIDATION_ERROR_CODES.AUTHORIZATION } },
+                }
+            );
+            const mappedError = mapError(error);
+            expect(mappedError).toHaveProperty("message");
+            expect(mappedError.message).toBe('Invalid argument: filter, error: Value "CREATE" does not exist in enum.');
+        });
```
Comment (darrellwarde) [inline packages/graphql/src/schema/validation/utils/map-error.test.ts:102]: This doesn't feel quite right, given that `AuthorizationValidateOperation` is a static enum which we could document?

File: packages/graphql/src/schema/validation/utils/map-error.test.ts [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+                    'Invalid argument: filter, error: Expected type "UserWhere" to be an object.',
+                    {
+                        nodes: [argumentNode, directiveNode],
+                        extensions: { exception: { code: VALIDATION_ERROR_CODES.AUTHORIZATION } },
+                    }
+                );
+                const mappedError = mapError(error);
+                expect(mappedError).toHaveProperty("message");
+                expect(mappedError.message).toBe("Invalid argument: filter, error: Expected type to be an object.");
+            });
```
Comment (darrellwarde) [inline packages/graphql/src/schema/validation/utils/map-error.test.ts:246]: Still not sure how I feel about types which will eventually exist in the schema...

File: packages/graphql/src/schema/validation/utils/map-error.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+const RENAMED_DIRECTIVE_OR_TYPE = /(\s?"([^\s]*?(Authorization|Authentication)[^\s]*?)")/;
+const WHERE_TYPE = /type(\s\\?".+?\\?")/; // <typename>Where / JwtPayloadWhere
+const JWT_PAYLOAD_DUMMY_VALUE_ERROR =
+    /(?:(?:String)|(?:Int)|(?:Float)|(?:Boolean))(?: cannot represent a?\s?)(?:(?:non string)|(?:non-integer)|(?:non numeric)|(?:non boolean))(?: value)(:.+)/;
+function composeRegex(predicate: string, ...regexes: RegExp[]) {
+    return new RegExp(regexes.map((r) => r.source).join(predicate));
+}
+function renameMysteryDirective(initialMessage: string): string | undefined {
```
Comment (darrellwarde) [inline packages/graphql/src/schema/validation/utils/map-error.ts:74]: "Mystery" is the word we want to go with here? 😆 Maybe worth a doc block on the function to explain the definition of a "mystery type" to us.

File: packages/graphql/src/schema/validation/schema-validation.test.ts [hunk 6]
```diff
@@ -450,75 +591,900 @@ describe("schema validation", () => {
...
+
+                const { typeDefs: augmentedDocument } = makeAugmentedSchema(userDocument);
+                const executeValidate = () => validateUserDefinition({ userDocument, augmentedDocument });
+
+                const errors = getError(executeValidate);
+                expect(errors).toHaveLength(1);
+                expect(errors[0]).not.toBeInstanceOf(NoErrorThrownError);
+                expect(errors[0]).toHaveProperty(
+                    "message",
+                    'Invalid argument: filter, error: Value "NEVER" does not exist in enum.'
```
Comment (darrellwarde) [inline packages/graphql/src/schema/validation/schema-validation.test.ts:654]: This is perhaps another example of when I maybe would have preserved the type name - up to you!

**Review by darrellwarde — COMMENTED** [review]
I think it looks pretty good! Just a few comments. 🙂

File: packages/graphql/src/schema/validation/utils/map-error.test.ts (hunk 3 above)
Comment (a-alle) [reply packages/graphql/src/schema/validation/utils/map-error.test.ts:102]: I wouldn't start discriminating between the mystery types and only erase some while keeping others. We could've gone all the way with documenting the validation types and used that approach for all..

File: packages/graphql/src/schema/validation/enrichers/directive/utils.ts (hunk 1 above)
Comment (a-alle) [reply packages/graphql/src/schema/validation/enrichers/directive/utils.ts:34]: They are passed to the enricher functions, for example the DIRECTIVE_TRANSFORM_FN  is passed as an argument to `directiveEnricher` which will call it with the encountered directive to transform it into the directive used as enrichment.
I thought it was better to have them explicitly defined although I see I forgot to use them in the enrichers/directive files..

Comment (a-alle) [reply packages/graphql/src/schema/validation/enrichers/directive/utils.ts:34]: usage updated

File: packages/graphql/src/schema/validation/utils/map-error.test.ts (hunk 3 above)
Comment (a-alle) [reply packages/graphql/src/schema/validation/utils/map-error.test.ts:102]: added some comments

File: packages/graphql/src/schema/validation/utils/map-error.ts (hunk 5 above)
Comment (a-alle) [reply packages/graphql/src/schema/validation/utils/map-error.ts:77]: added some comments

**Pull request merged**